src/level_cache.cpp

```diff
@@ -0,0 +1,107 @@
+#include "level_cache.h"
+
+#include <cstdlib>
+
+#include "vehicle.h"
+
+level_cache::level_cache()
+{
+    const int map_dimensions = MAPSIZE_X * MAPSIZE_Y;
+    transparency_cache_dirty.set();
+    outside_cache_dirty = true;
+    floor_cache_dirty = false;
+    constexpr four_quadrants four_zeros( 0.0f );
+    std::fill_n( &lm[0][0], map_dimensions, four_zeros );
+    std::fill_n( &sm[0][0], map_dimensions, 0.0f );
+    std::fill_n( &light_source_buffer[0][0], map_dimensions, 0.0f );
+    std::fill_n( &outside_cache[0][0], map_dimensions, false );
+    std::fill_n( &floor_cache[0][0], map_dimensions, false );
+    std::fill_n( &transparency_cache[0][0], map_dimensions, 0.0f );
+    std::fill_n( &vision_transparency_cache[0][0], map_dimensions, 0.0f );
+    std::fill_n( &seen_cache[0][0], map_dimensions, 0.0f );
+    std::fill_n( &camera_cache[0][0], map_dimensions, 0.0f );
+    std::fill_n( &visibility_cache[0][0], map_dimensions, lit_level::DARK );
+    veh_in_active_range = false;
+    std::fill_n( &veh_exists_at[0][0], map_dimensions, false );
+}
+
+bool level_cache::get_veh_in_active_range() const
+{
+    return veh_in_active_range;
+}
+
+bool level_cache::get_veh_exists_at( const tripoint &pt ) const
+{
+    return veh_exists_at[ pt.x ][ pt.y ];
+}
+
+std::pair<vehicle *, int> level_cache::get_veh_cached_parts( const tripoint &pt ) const
+{
+    auto it = veh_cached_parts.find( pt );
+    if( it != veh_cached_parts.end() ) {
+        return it->second;
+    }
+    vehicle *veh = nullptr;
+    return std::make_pair( veh, -1 );
+}
+
+void level_cache::set_veh_in_active_range( bool is_active )
+{
+    veh_in_active_range = is_active;
+    veh_cache_dirty = true;
+    veh_cache_active |= is_active;
+}
+
+void level_cache::set_veh_exists_at( const tripoint &pt, bool exists_at )
+{
+    veh_exists_at[ pt.x ][ pt.y ] = exists_at;
+    veh_cache_dirty = true;
+    veh_cache_active |= exists_at;
+}
+
+void level_cache::set_veh_cached_parts( const tripoint &pt, vehicle *veh, int part_num )
```

Comment:
Is `veh` arg here ever expected to be `nullptr`? 

If yes, then what's the point of `veh_cached_parts[ pt ] = std::make_pair( veh, part_num );` when `veh == nullptr`?
If `veh` cannot be null, `veh_cache_active |= veh != nullptr;` doesn't make sense.

I suspect that `veh` is not expected to be `nullptr` and if that's the case, it probably makes sense to change the type of the `veh` argument to `const vehicle &veh`.

Reply:
so the veh_cached_parts is returned through `veh_at_internal()` as `vehicle *`; I'm not going to change that and I'm nervous about changing the stored `vehicle *` but I guess I can pass it as a reference.

Reply:
I think changing the method signature to `void set_veh_cached_parts( const tripoint &pt, const vehicle &veh, int part_num )` and replacing `veh_cache_active |= veh != nullptr;` inside with `veh_cache_active = true;` would make it a bit cleaner, but it's also ok if you leave it as is.

Reply:
Fixed, thanks for the suggestion.